### ISSUE TITLE
deprecation of log_group_name in aws_vpc_flow_log change breaks aws_cloudwatch_log_subscription_filter semantics

### ISSUE THREAD

**Issue by bacoboy** [terraform-providers/terraform-provider-aws#6360]
Using the current aws provider (1.42.0 at time of writing), a change in `aws_vpc_flow_logs` replaces the `log_group_name` string with a `log_destination` ARN.  This, I believe was part of adding support for S3 targets in cloudwatch logs.

Until now, the `aws_vpc_flow_logs.NAME.log_group_name` was used directly as input for the `aws_cloudwatch_log_subscription_filter`'s `log_group_name` field.

While the field still exists, it has a `:*` appended to the end (when you use the `log_destination` version, which now makes it invalid as input. So this code:
```
resource "aws_cloudwatch_log_subscription_filter" "shipper" {
  depends_on = ["aws_lambda_permission.shipper"]
  name = "shipper"
  log_group_name = "${aws_flow_log.vpc.log_group_name}"
  filter_pattern = "[version, account_id, interface_id, srcaddr != \"-\", dstaddr != \"-\", srcport != \"-\", dstport != \"-\", protocol, packets, bytes, start, end, action, log_status]"
  destination_arn = "${aws_lambda_function.shipper.arn}"
}

```
will fail:
```
* aws_cloudwatch_log_subscription_filter.shipper: InvalidParameterException: 1 validation error detected: Value '/vpc_flow_logs/main:*' at 'logGroupName' failed to satisfy constraint: Member must satisfy regular expression pattern: [\.\-_/#A-Za-z0-9]+
	status code: 400, request id: d5855e7c-e10a-11e8-adae-8ddab33fd8a5
```
A plan shows why.  The log_group_name has a `:*` at the end:
```
  + aws_cloudwatch_log_subscription_filter.shipper
      id:              <computed>
      destination_arn: "arn:aws:lambda:us-east-1:123456789012:function:shipper"
      filter_pattern:  "[version, account_id, interface_id, srcaddr != \"-\", dstaddr != \"-\", srcport != \"-\", dstport != \"-\", protocol, packets, bytes, start, end, action, log_status]"
      log_group_name:  "/vpc_flow_logs/main:*"
      name:            "shipper"
      role_arn:        <computed>
```

A workaround is to do something like this:
```
  log_group_name = "${element(split(":", aws_flow_log.NAME.log_group_name), 0)}"
```
but this is awful and `log_group_name` will eventually go away.

Here is a complete example you can run.  The lambda `shipper.js`:
```
exports.handler = async (event, context) => {
    console.log('Received event:', JSON.stringify(event, null, 2));
    return event;
};
```
And the full terraform code:
```
provider "aws" {
  region = "us-east-1"
  profile = "your-admin-profile"
}

variable "aws_region" { default = "us-east-1" }
resource "aws_vpc" "main" {
  cidr_block = "192.168.0.0/16"
}

resource "aws_cloudwatch_log_group" "vpc" {
  name = "/vpc_flow_logs/main"
  retention_in_days = 1
}

resource "aws_iam_role" "shipper" {
    name = "shipper"
    assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Service": [
          "logs.${var.aws_region}.amazonaws.com",
          "lambda.amazonaws.com"
        ]
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
EOF
}

resource "aws_iam_role_policy" "shipper" {
  name = "shipper"
  role = "${aws_iam_role.shipper.id}"
  policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "iam:PassRole",
      "Resource": "${aws_iam_role.shipper.arn}"
    },
    {
      "Effect": "Allow",
      "Action": [
          "logs:CreateLogGroup",
          "logs:CreateLogStream",
          "logs:PutLogEvents"
      ],
      "Resource": "${aws_cloudwatch_log_group.vpc.arn}"
    }
  ]
}
EOF
}

resource "aws_flow_log" "vpc" {
    iam_role_arn = "${aws_iam_role.shipper.arn}"
    log_destination_type = "cloud-watch-logs"
    log_destination =  "${aws_cloudwatch_log_group.vpc.arn}"
    traffic_type = "ALL"
    vpc_id = "${aws_vpc.main.id}"
}

data "archive_file" "shipper" {
    type        = "zip"
    source_file  = "shipper.js"
    output_path = "shipper.zip"
}
resource "aws_lambda_function" "shipper" {
  filename         = "shipper.zip"
  function_name    = "shipper"
  description      = "Log shipping lambda"
  role             = "${aws_iam_role.shipper.arn}"
  handler          = "exports.handler"
  source_code_hash = "${data.archive_file.shipper.output_base64sha256}"
  runtime          = "nodejs8.10"
  memory_size      = "128"
  timeout          = "5"
}
resource "aws_cloudwatch_log_group" "shipper" {
  name = "/aws/lambda/${aws_lambda_function.shipper.function_name}"
  retention_in_days = 1
}

# grant cloudwatch logs ability to invoke (needed for subscription)
resource "aws_lambda_permission" "shipper" {
  statement_id   = "AllowExecutionFromCloudWatch"
  function_name  = "${aws_lambda_function.shipper.function_name}"
  action         = "lambda:InvokeFunction"
  principal      = "logs.${var.aws_region}.amazonaws.com"
  source_arn     = "${aws_cloudwatch_log_group.vpc.arn}"
}

# Setup Subscription from log group
resource "aws_cloudwatch_log_subscription_filter" "shipper" {
  depends_on = ["aws_lambda_permission.shipper"]
  name = "shipper"
  # They added a :* at the end of the name (plus log_group_name in flow logs due to s3 addition is deprecated so expect a change here too)
  # This is the workaround:
  #log_group_name = "${element(split(":", aws_flow_log.vpc.log_group_name), 0)}"
  log_group_name = "${aws_flow_log.vpc.log_group_name}"
  filter_pattern = "[version, account_id, interface_id, srcaddr != \"-\", dstaddr != \"-\", srcport != \"-\", dstport != \"-\", protocol, packets, bytes, start, end, action, log_status]"
  destination_arn = "${aws_lambda_function.shipper.arn}"
}
```

**Comment by bflad** [comment]
I believe this problem stems from `aws_cloudwatch_log_group` `arn` coming back from the API with `:*` appended (without our doing: https://github.com/terraform-providers/terraform-provider-aws/blob/4148c39b055ecc393229b8ee3b2f350aa2b371fb/aws/resource_aws_cloudwatch_log_group.go#L115)

Unfortunately to fix that (by creating a new attribute to keep the `:*` and removing `:*` from the existing `arn` attribute) will be a breaking change.

**Comment by bacoboy** [comment]
Well the `log_group_name` on the flow log has been deprecated.  Seems either than needs to become an output (without the `:*` to be consumed by a subscription filter) or the subscription filter needs to follow suit and switch to a log_group ARN as input.

**Comment by bflad** [comment]
The above change has been merged and will release with version 1.43.0 of the AWS provider, likely later today or tomorrow. 👍

**Comment by bflad** [comment]
This has been released in [version 1.43.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1430-november-07-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
